core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,357 @@
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> successfulTasks = new HashSet<>();
+    @GuardedBy("this")
+    private final Map<TaskId, Throwable> failedTasks = new HashMap<>();
+    @GuardedBy("this")
+    private boolean inputFinished;
+    @GuardedBy("this")
+    private Throwable failure;
+
+    @GuardedBy("this")
+    private final ListMultimap<TaskId, SerializedPage> pageBuffer = LinkedListMultimap.create();
+    @GuardedBy("this")
+    private Iterator<SerializedPage> pagesIterator;
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private int maxAttemptId;
+
+    @GuardedBy("this")
+    private boolean closed;
+
+    public DeduplicationExchangeClientBuffer(Executor executor, DataSize bufferCapacity, RetryPolicy retryPolicy)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+        requireNonNull(retryPolicy, "retryPolicy is null");
+        checkArgument(retryPolicy == QUERY, "retryPolicy is expected to be QUERY: %s", retryPolicy);
+        this.retryPolicy = retryPolicy;
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
+    }
+
+    @Override
+    public synchronized SerializedPage pollPage()
+    {
+        throwIfFailed();
+
+        if (closed) {
+            return null;
+        }
+
+        if (!inputFinished) {
+            return null;
+        }
+
+        if (pagesIterator == null) {
+            pagesIterator = pageBuffer.values().iterator();
+        }
+
+        if (!pagesIterator.hasNext()) {
+            return null;
+        }
+
+        SerializedPage page = pagesIterator.next();
+        pagesIterator.remove();
+        bufferRetainedSizeInBytes -= page.getRetainedSizeInBytes();
+
+        return page;
+    }
+
+    @Override
+    public synchronized void addTask(TaskId taskId)
+    {
+        if (closed) {
+            return;
+        }
+
+        checkState(!noMoreTasks, "no more tasks expected");
+        checkState(allTasks.add(taskId), "task already registered: %s", taskId);
+
+        if (retryPolicy == QUERY) {
+            removePagesForPreviousAttempts(taskId.getAttemptId());
+        }
+
+        if (taskId.getAttemptId() > maxAttemptId) {
+            maxAttemptId = taskId.getAttemptId();
+        }
```

Comment:
This will remove unnecessary traversing of `pageBuffer`, right?
```suggestion
        if (taskId.getAttemptId() > maxAttemptId) {
            maxAttemptId = taskId.getAttemptId();
            
            if (retryPolicy == QUERY) {
              removePagesForPreviousAttempts(maxAttemptId);
          }
        }
```

Reply:
Good catch. Let me update